src/main.rs

```diff
@@ -69,16 +76,18 @@ fn main() {
     // Parse & Build an AST
     let mut parser = Parser::new(lexer);
 
-    let ast_root = match parser.parse() {
+    let mut ast_root = match parser.parse() {
```

Comment:
It shouldnt be any more work, maybe transiently more memory (though if you were to generate a structure that has references to points on the AST tree, rather than duplicating the information, it might not even be that much), but its totally a thing. AST is a direct map from the parser, an IL doesn't necessarily, and has more relevant information for codegen and typing and whatnot, whereas the AST wont have any of that.

It also means you don't have to create this structure inside the parser that has a bunch of information (type and blah and whatnot that you generate through typechecking and scope checking and whatever) that's just zeroed out because you can't yet generate it.

It's probably not 100% necessary at this point, but I'd figure it would be easier to do earlier than later.